lib/sendgrid/mail.rb

```diff
@@ -28,34 +33,46 @@ def add_attachment(path, name = nil)
     end
 
     def to_h
-      payload = {
-        :from        => @from,
-        :fromname    => @from_name,
-        :subject     => @subject,
-        :to          => @to,
-        :toname      => @to_name,
-        :date        => @date,
-        :replyto     => @reply_to,
-        :cc          => @cc,
-        :bcc         => @bcc,
-        :text        => @text,
-        :html        => @html,
+      payload = {}
+
+      PAYLOAD_PARAMS.each do |payload_param|
+        payload[payload_mapping(payload_param)] = self.send(payload_param)
+      end
+
+      payload.merge!({
         :'x-smtpapi' => smtpapi_json,
         :files       => ({} unless @attachments.empty?)
-      }.reject {|k,v| v.nil?}
+      })
+
+      payload.reject! {|k,v| v.nil?}
+
+      assign_missing_to(payload)
+      extract_attachments(payload)
```

Comment:
There's a code smell here.

In the above `payload.merge!` call the `:files` key can be `nil`, however `extract_attachments` can fail if it is `nil`. Currently they both check `@attachments.empty?` so nothing bad should happen, but I think we can do better.

What about place thing directly into the `merge!` call
```ruby
:files => extracted_attachments
```
And write `extracted_attachments` to return `nil` or the extracted attachment hash.